### PR TITLE
Fix i64 select lowering.

### LINKED ISSUE #1767: wasm2js miscompiles select instruction with i64 operands
It seems that when `select` instructions with i64 operands are compiled to js only the low bits are actually set according to the condition.

The following code
```wast
(module
 (type $0 (func (result i64)))
 (func $0 (; 0 ;) (type $0) (result i64)
  (return
   (select
    (i64.const -1)
    (i64.const 0)
    (i32.const 1)
   )   
  )    
 )     
)
```

compiles to
```js
 var i64toi32_i32$HIGH_BITS = 0;
 function $0() {
  var i64toi32_i32$1 = 0, i64toi32_i32$0 = 0, wasm2js_i32$0 = 0, wasm2js_i32$1 = 0, wasm2js_i32$2 = 0;
  i64toi32_i32$1 = 0;
  i64toi32_i32$0 = i64toi32_i32$1;
  i64toi32_i32$1 = (wasm2js_i32$0 = 4294967295, wasm2js_i32$1 = 0, wasm2js_i32$2 = 1, wasm2js_i32$2 ? wasm2js_i32$0 : wasm2js_i32$1);
  i64toi32_i32$HIGH_BITS = i64toi32_i32$0;
  return i64toi32_i32$1 | 0;
 }
```
which gives the correct answer (-1) in the lower bits but the high bits are incorrectly set to 0.

### PR DESCRIPTION
**Pull request by yurydelendik**
Fixes #1767

### CODE REVIEW

File: src/passes/I64ToI32Lowering.cpp [hunk 1]
```diff
@@ -1526,7 +1520,28 @@ struct I64ToI32Lowering : public WalkerPass<PostWalker<I64ToI32Lowering>> {
   }
 
   void visitSelect(Select* curr) {
-    visitBranching<Select>(curr);
+    if (!hasOutParam(curr->ifTrue)) return;
```
Comment (kripken) [inline src/passes/I64ToI32Lowering.cpp]: Please add an assert that neither does the `ifFalse` in that case

File: src/passes/I64ToI32Lowering.cpp [hunk 2]
```diff
@@ -1526,7 +1520,28 @@ struct I64ToI32Lowering : public WalkerPass<PostWalker<I64ToI32Lowering>> {
...
+    TempVar trueBits = fetchOutParam(curr->ifTrue);
+    TempVar highBits = fetchOutParam(curr->ifFalse);
```
Comment (kripken) [inline src/passes/I64ToI32Lowering.cpp]: perhaps `falseBits` to match the line above?

File: src/passes/I64ToI32Lowering.cpp [hunk 3]
```diff
@@ -1526,7 +1520,28 @@ struct I64ToI32Lowering : public WalkerPass<PostWalker<I64ToI32Lowering>> {
...
+            highBits,
+            builder->makeGetLocal(trueBits, i32)
+          ),
+          builder->makeGetLocal(tmpTrue, i32)
+        ),
+        builder->makeGetLocal(tmpFalse, i32)
+      )
+    );
+    replaceCurrent(result);
+    setOutParam(result, std::move(highBits));
```
Comment (kripken) [inline src/passes/I64ToI32Lowering.cpp]: I may be missing something, but shouldn't the out param be a select over either the high bits of the ifTrue or the ifFalse?

Comment (yurydelendik) [reply src/passes/I64ToI32Lowering.cpp]: Sorry, highBits (now falseBits) will contain that. I got carried away with optimizations and pushed wrong commit. Changed makeIf to makeSelect, reverted this.

File: src/passes/I64ToI32Lowering.cpp (hunk 2 above)
Comment (yurydelendik) [reply src/passes/I64ToI32Lowering.cpp]: done

File: src/passes/I64ToI32Lowering.cpp (hunk 1 above)
Comment (yurydelendik) [reply src/passes/I64ToI32Lowering.cpp]: Fixed

File: src/passes/I64ToI32Lowering.cpp [hunk 4]
```diff
@@ -1526,7 +1520,29 @@ struct I64ToI32Lowering : public WalkerPass<PostWalker<I64ToI32Lowering>> {
...
+    if (!hasOutParam(curr->ifTrue)) return;
+    assert(hasOutParam(curr->ifFalse) && "hasOutParam(ifFalse) is false");
+    TempVar trueBits = fetchOutParam(curr->ifTrue);
+    TempVar falseBits = fetchOutParam(curr->ifFalse);
+    TempVar tmpTrue = getTemp();
+    TempVar tmpFalse = getTemp();
+    Block* result = builder->blockify(
+      builder->makeSetLocal(tmpTrue, curr->ifTrue),
+      builder->makeSetLocal(tmpFalse, curr->ifFalse),
+      builder->makeIf(
```
[the remark this answers is not in the document]
Comment (yurydelendik) [reply src/passes/I64ToI32Lowering.cpp]: Yeah, I tried. With `select` we are doing the same mistake/bug (but for 'true' branch in this case). Also, I tried to simplify logic and lower amount of temps without any luck so far.

[the remark this answers is not in the document]
Comment (yurydelendik) [reply src/passes/I64ToI32Lowering.cpp]: One of my solutions was:
```
    TempVar highBits = getTemp();
    TempVar cond = getTemp();
    Block* result = builder->blockify(
      builder->makeSetLocal(
        highBits,
        builder->makeSelect(
          builder->makeTeeLocal(cond, curr->condition),
          builder->makeGetLocal(fetchOutParam(curr->ifTrue), i32),
          builder->makeGetLocal(fetchOutParam(curr->ifFalse), i32)
        )
      ),
      builder->makeSelect(
        builder->makeGetLocal(cond, i32),
        curr->ifTrue,
        curr->ifFalse
      )
    );
    replaceCurrent(result);
    setOutParam(result, std::move(highBits));
```

that produced for the second function in the i64-select.wast the following:

```
  var i64toi32_i32$3 = 0, i64toi32_i32$2 = 0, i64toi32_i32$0 = 0, i64toi32_i32$1 = 0, wasm2js_i32$0 = 0, wasm2js_i32$1 = 0, wasm2js_i32$2 = 0;
  i64toi32_i32$3 = 1;
  i64toi32_i32$2 = (wasm2js_i32$0 = i64toi32_i32$0, wasm2js_i32$1 = i64toi32_i32$1, wasm2js_i32$2 = i64toi32_i32$3, wasm2js_i32$2 ? wasm2js_i32$0 : wasm2js_i32$1);
  i64toi32_i32$0 = 4294967295; // ??? why i64toi32_i32$0 initialized after i64toi32_i32$2
  i64toi32_i32$1 = 0;
  i64toi32_i32$2 = i64toi32_i32$2;
  i64toi32_i32$3 = (wasm2js_i32$0 = 4294967295, wasm2js_i32$1 = 0, wasm2js_i32$2 = i64toi32_i32$3, wasm2js_i32$2 ? wasm2js_i32$0 : wasm2js_i32$1);
  i64toi32_i32$HIGH_BITS = i64toi32_i32$2;
  return i64toi32_i32$3 | 0;
```

**Comment by ngg** [comment]
lgtm, works well for my original testcase from which I created that small reproduction case, thanks!

File: src/passes/I64ToI32Lowering.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/I64ToI32Lowering.cpp]: The issue may be the tee_local. We may depend on the IR being flat, and tee violates that. Doing the set_local before, and a get for the select condition, should work (if not, let's look more into why, I can help debug).

[the remark this answers is not in the document]
Comment (yurydelendik) [reply src/passes/I64ToI32Lowering.cpp]: the same problem was with set_local as well. I swapped order of calculation for low and highBits -- now it works as expected.

File: src/passes/I64ToI32Lowering.cpp [hunk 5]
```diff
@@ -1526,7 +1520,35 @@ struct I64ToI32Lowering : public WalkerPass<PostWalker<I64ToI32Lowering>> {
...
+    }
+    assert(hasOutParam(curr->ifFalse));
+    TempVar highBits = getTemp();
+    TempVar lowBits = getTemp();
+    TempVar cond = getTemp();
+    Block* result = builder->blockify(
+      builder->makeSetLocal(
+        lowBits,
+        builder->makeSelect(
+          builder->makeTeeLocal(cond, curr->condition),
```
Comment (kripken) [inline src/passes/I64ToI32Lowering.cpp]: this tee violates the flat property of the IR. I think we should preserve it even if we don't depend on that right now. That means just doing a set earlier in the block being created here, and a get here.

**Pull request merged**